README.md
# PageLocks Plugin for [Grav](http://getgrav.org)

PageLocks is a useful plugin for Grav installations with several users working in the Admin backend. It activates a lock for any page that someone is editing so that nobody else can accidentally edit the same page (but they can look at it).

This only works if all editors are using the Admin panel. It will NOT help if
- someone uploads changes via FTP
- someone is using one of the frontend editing plugins.

## Many many thanks go to
[pamtbaau](https://github.com/pamtbaau) who whipped **all** this code up without breaking a sweat, just because I'd asked in the forum how I might achieve such a pagelock thing. I'm really just fixing up the details here, humbly vowing to improve my coding so I can at least maintain all of this and solve issues by myself. ALL credits go to them!

## Installation

Installing the PageLocks plugin can be done in one of three ways: The GPM (Grav Package Manager) installation method lets you quickly install the plugin with a simple terminal command, the manual method lets you do so via a zip file, and the admin method lets you do so via the Admin Plugin.

### Admin Plugin (easiest)

If you use the Admin Plugin (which you most certainly do because otherwise you don't need this plugin), you can install PageLocks directly by browsing the `Plugins`-menu and clicking on the `Add` button.

### GPM Installation

To install the plugin via the [GPM](http://learn.getgrav.org/advanced/grav-gpm), through your system’s terminal (also called the command line), navigate to the root of your Grav-installation, and enter:

    bin/gpm install automagic-images

This will install the PageLocks plugin into your `/user/plugins`-directory within Grav. Its files can be found under `/your/site/grav/user/plugins/pagelocks`.

### Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `pagelocks`. You can find these files on [GitHub](https://github.com/skinofthesoul/grav-plugin-pagelocks) or via [GetGrav.org](http://getgrav.org/downloads/plugins).

You should now have all the plugin files under

    /your/site/grav/user/plugins/pagelocks


## Configuration
Here is the default configuration and an explanation of available options:

```
enabled: true
expiresAfter: 60            # Lock expires after x seconds of no keepAlive pings
keepAliveInterval: 30       # Send every x seconds a keepAlive ping to server
productionMode: false       # Add minified assets if true
debug: true                 # If true, write a log to data/pagelocks/debug.log
```

## How does it work?
If you'd like to see the plugin in action yourself, get two different user logins for your Grav installation and log into Admin with each of them in a separate browser, then try to edit the same page with both.

### Features:
- When lock is acquired, user is prevented from editing the page:
  - The form is "blocked" by adding an extra layer on top.
  - Editing buttons Move, Delete and Save are hidden
  - User can no longer toggle between Normal|Expert mode.
- Locks expire after `expiresAfter` seconds if no keepAlive has arrived during that period. 
  - E.g. when user moved out of Admin, or user went out for lunch...
  - User is notified when lock has been forcefully removed, or has expired.
- Keep alive request are being sent every `keepAliveInterval` seconds.
- Using option `debug` all lock state changes can me logged
- Messages in client are translated
- Users are notified in banner in top of page
- Javascript is generated using strongly typed Typescript


### Frontend: in Admin
- Two javascript files are added to Admin panel
  - `pagelocker.js`: Injected into every Admin page
    - Sends async requests to acquire lock
    - Sends keepAlive requests to prevent lock from expiring
  - `pagelocksadmin.js`: Injected only into Admin page `/admin/locks`, the page to list/remove locks
    - Lists all acquired locks for all pages/users
    - User can sends async request to remove lock for a certain route

### Backend
- All locks are kept in file `/data/pagelocks/locks.yaml`
- On each async request on the server, any expired lock is removed
- If route of page is like `/admin/pages(/pagename)+`:
  - Any existing lock of current user is removed
  - A new lock is acquired for user
- Else:
  - Any existing lock of current user is removed

## Todo:
- [ ] test & improve: When a lock is forcibly removed, but the user still has the page open, it can take a while for them to get the "lock removed" notice.
- [ ] add minification for CSS


CHANGELOG.md
# v0.1.3
##  08-07-2021

1. [](#bugfix)
    * allow for different admin route
2. [](#improved)
    * made headlines in admin list of pagelocks translatable

# v0.1.2
##  08-07-2021

1. [](#bugfix)
    * vendor files added

# v0.1.1
##  08-07-2021

1. [](#improved)
    * added German language support and some config settings in Admin
    * fixed some tiny details (typos, icons and the like)

# v0.1.0
##  14-06-2021

1. [](#new)
    * initial draft & commit by pamtbaau


tests/unit/LockHandlerTest.php
<?php

use Grav\Common\Grav;
use Grav\Common\User\DataUser\User;
use Grav\Plugin\PageLocks\LockHandler;
use RocketTheme\Toolbox\File\YamlFile;
use RocketTheme\Toolbox\ResourceLocator\UniformResourceLocator;

class LockHandlerTest extends \Codeception\Test\Unit
{
    const LOCKFILE = 'user-data://pagelocks/locks.yaml';

    protected YamlFile $lockFile;

    public function _before()
    {
        /** @var Grav */
        $grav = Grav::instance();

        /** @var User */
        $user = new User();
        $user->email = 'user1@domain.com';
        $user->fullname = 'User 1';
        $user->save();

        $grav['user'] = $user;

        /** @var UniformResourceLocator */
        $locator = $grav['locator'];
        $lockPath = $locator->findResource(self::LOCKFILE, true, true);

        if ($lockPath === false) {
            $this->fail('Path to Lockfile could not be found.');
        }

        $this->lockFile = YamlFile::instance($lockPath);
    }

    protected function _after(): void
    {
    }

    public function testReadEmptyLocks(): void
    {
        $this->set_POST([
            'readLocks' => '',
        ]);
        $this->set_LOCKFILE([]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(0, count($response['locks']));
        $this->assertEquals('Have you confirmed the page is no longer being edited by %s?', $response['alert']);
        $this->assertEquals('Found 0 lock(s)', $response['countAlert']);
    }

    public function testReadLocks(): void
    {
        $this->set_POST([
            'readLocks' => '',
        ]);

        $this->set_LOCKFILE([
            '/admin/pages/page1' => [
                'email' => 'user1@domain.com',
                'fullname' => 'User 1',
                'timestamp' => 1234556789,
            ]
        ]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(1, count($response['locks']));
        $this->assertEquals(true, isset($response['locks']['/admin/pages/page1']));
        $this->assertEquals('user1@domain.com', $response['locks']['/admin/pages/page1']['email']);
        $this->assertEquals('Have you confirmed the page is no longer being edited by %s?', $response['alert']);
        $this->assertEquals('Found 1 lock(s)', $response['countAlert']);
    }

    public function testAcquireLockAlreadyLocked(): void
    {
        $this->set_POST([
            'acquireLock' => '/admin/pages/page1',
        ]);

        $this->set_LOCKFILE([
            '/admin/pages/page1' => [
                'email' => 'user2@domain.com',
                'fullname' => 'User 2',
                'timestamp' => 1234556789,
            ]
        ]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(true, $response['isOnPage']);
        $this->assertEquals(false, $response['isLockGranted']);
        $this->assertEquals('User 2', $response['byUser']);
        $this->assertEquals(60000, $response['keepAliveInterval']);
        $this->assertEquals("User 2 is currently editing this page. \nPlease try again later, or contact User 2 to coordinate editing of page.\n", $response['alert']);
    }

    public function testAcquireLockGranted(): void
    {
        $this->set_POST([
            'acquireLock' => '/admin/pages/page2',
        ]);

        $this->set_LOCKFILE([
            '/admin/pages/page1' => [
                'email' => 'user2@domain.com',
                'fullname' => 'User 2',
                'timestamp' => 1234556789,
            ]
        ]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(true, $response['isOnPage']);
        $this->assertEquals(true, $response['isLockGranted']);
        $this->assertEquals('User 1', $response['byUser']);
        $this->assertEquals(60000, $response['keepAliveInterval']);
        $this->assertEquals('Locks has been acquired successfully', $response['alert']);
    }

    public function testClearExpiredLocks(): void
    {
        $this->set_POST([
            'acquireLock' => '/admin/pages/page2',
        ]);

        $this->set_LOCKFILE([
            '/admin/pages/page1' => [
                'email' => 'user2@domain.com',
                'fullname' => 'User 2',
                'timestamp' => 0,
            ]
        ]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        $this->set_POST([
            'readLocks' => '',
        ]);

        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(1, count($response['locks']));
        $this->assertEquals(true, isset($response['locks']['/admin/pages/page2']));
    }

    public function testForceRemoveLock(): void
    {
        $this->set_POST([
            'forceRemoveLock' => '/admin/pages/page2',
        ]);

        $this->set_LOCKFILE([
            '/admin/pages/page1' => [
                'email' => 'user1@domain.com',
                'fullname' => 'User 1',
                'timestamp' => 0,
            ],
            '/admin/pages/page2' => [
                'email' => 'user2@domain.com',
                'fullname' => 'User 2',
                'timestamp' => 0,
            ]
        ]);

        $lockHandler = new LockHandler();
        $response = $lockHandler->handleRequest();

        if ($response === null) {
            $this->fail('Variable $response is null');
        }

        $this->assertEquals(true, $response['isLockRemoved']);
        $this->assertEquals('Lock has been removed successfully.', $response['alert']);
    }

    private function set_LOCKFILE(array $locks): void
    {
        $this->lockFile->save($locks);
        $this->lockFile->free();
    }

    private function set_POST(array $post): void
    {
        global $_POST;
        $_POST = $post;
    }
}


tests/acceptance/PageLocksCest.php
<?php

use Grav\Common\Grav;
use RocketTheme\Toolbox\File\File;
use RocketTheme\Toolbox\File\YamlFile;
use RocketTheme\Toolbox\ResourceLocator\UniformResourceLocator;

use function PHPUnit\Framework\assertArrayHasKey;
use function PHPUnit\Framework\assertArrayNotHasKey;
use function PHPUnit\Framework\assertCount;
use function PHPUnit\Framework\assertTrue;

/**
 * - Dummy Author 2 is granted a dummy lock in the lock file on page /admin/pages/typography
 * - Auhtor 1 is logging in and tries te get locks on home (succeeds) and typography (fails)
 */

class PageLocksCest
{
    const LOCKFILE = 'user-data://pagelocks/locks.yaml';
    const LOGFILE = 'user-data://pagelocks/debug.log';
    const CONFIGFILE = 'config://plugins/pagelocks.yaml';

    protected YamlFile $lockFile;
    protected YamlFile $configFile;
    protected File $logFile;

    protected array $configDefaults = [
        'enabled' => true,
        'expiresAfter' =>  3600,
        'keepAliveInterval' =>  60,
        'productionMode' =>  true,
        'debug' =>  true,
    ];

    public function __construct()
    {
        $grav = Grav::instance();

        /** @var UniformResourceLocator */
        $locator = $grav['locator'];

        $lockPath = $locator->findResource(self::LOCKFILE, true, true);
        if ($lockPath === false) {
            throw new Exception('Resource "self::LOCKFILE" could not be found');
        }
        $this->lockFile = YamlFile::instance($lockPath);

        $configPath = $locator->findResource(self::CONFIGFILE, true, true);
        if ($configPath === false) {
            throw new Exception('Resource "self::CONFIGFILE" could not be found');
        }
        $this->configFile = YamlFile::instance($configPath);

        $logPath = $locator->findResource(self::LOGFILE, true, true);
        
        if ($logPath === false) {
            throw new Exception('Resource "self::LOGFILE" could not be found');
        }
        $this->logFile = File::instance($logPath);

        $this->logFile->save('');
        $this->logFile->free();

        $this->setConfigDefaults();
    }

    public function _before(): void
    {
    }

    public function _after(): void
    {
        $this->setConfigDefaults();
        $this->set_LOCKFILE([]);
    }

    public function grantLock(AcceptanceTester $I): void
    {
        // Add dummy lock for Author2 on page /admin/pages/typography
        $this->set_LOCKFILE([
            '/admin/pages/typography' => [
                'email' => 'author2@domain.com',
                'fullname' => 'McAuthor 2',
                'timestamp' => time(),
            ]
        ]);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/home');
        $I->see('Home');

        $locks = $this->get_LOCKFILE();

        assertCount(2, $locks);
        assertArrayHasKey('/admin/pages/home', $locks);
        assertArrayHasKey('/admin/pages/typography', $locks);
    }

    public function denyLock(AcceptanceTester $I): void
    {
        // Add dummy lock for Author2 on page /admin/pages/typography
        $this->set_LOCKFILE([
            '/admin/pages/typography' => [
                'email' => 'author2@domain.com',
                'fullname' => 'McAuthor 2',
                'timestamp' => time(),
            ]
        ]);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/typography');
        $I->see('Typography');
        // Check if Lock error alert is shown
        $I->seeElement('.error.alert.pagelocks');
        $I->see('author2@domain.com is currently editing this page.');

        $locks = $this->get_LOCKFILE();

        assertCount(1, $locks);
        assertArrayHasKey('/admin/pages/typography', $locks);
    }

    public function checkLocksAfterPageSwitch(AcceptanceTester $I): void
    {
        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/home');
        $I->see('Home');

        $locks = $this->get_LOCKFILE();

        assertCount(1, $locks);
        assertArrayHasKey('/admin/pages/home', $locks);

        // Switch to Typography

        $I->amOnPage('/admin/pages/typography');
        $I->see('Typography');

        $locks = $this->get_LOCKFILE();

        assertCount(1, $locks);
        assertArrayHasKey('/admin/pages/typography', $locks);
    }

    public function clearExpiredLocks(AcceptanceTester $I): void
    {
        // Add dummy lock for Author2 on page /admin/pages/typography
        $this->set_LOCKFILE([
            '/admin/pages/dummy1' => [
                'email' => 'dummy1@domain.com',
                'fullname' => 'Dummy 1',
                'timestamp' => 0,     // Timestamp of 1970-01-01
            ],
            '/admin/pages/dummy2' => [
                'email' => 'dummy2@domain.com',
                'fullname' => 'Dummy 2',
                'timestamp' => 0,    // Timestamp of 1970-01-01
            ]
        ]);

        $locks = $this->get_LOCKFILE();
        assertCount(2, $locks);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/home');
        $I->see('Home');

        $locks = $this->get_LOCKFILE();
        assertCount(1, $locks);
    }

    public function keepAlive(AcceptanceTester $I): void
    {
        $this->setConfig('keepAliveInterval', 1);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/home');
        $I->see('Home');

        $I->wait(5);

        $log = $this->logFile->content();
        $extendsFound = preg_match_all('/Lock extended: Held by author1@domain.com/', $log);

        assertTrue($extendsFound > 4);
    }

    public function forceRemoveLock(AcceptanceTester $I): void
    {
        // Add dummy lock for Author2 on page /admin/pages/typography
        $this->set_LOCKFILE([
            '/admin/pages/typography' => [
                'email' => 'author2@domain.com',
                'fullname' => 'McAuthor 2',
                'timestamp' => time(),
            ]
        ]);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/locks');
        $I->see('Locks');

        $I->seeElement('.page-delete.delete-action');
        $I->click('.page-delete.delete-action');

        $I->acceptPopup();

        $I->waitForText('Lock has been removed successfully.');
        $I->see('Lock has been removed successfully.');

        $I->waitForText('Found 0 lock(s)');
        $I->see('Found 0 lock(s)');

        $locks = $this->get_LOCKFILE();
        assertCount(0, $locks);
        assertArrayNotHasKey('/admin/pages/typography', $locks);
    }

    public function checkInactiveUserLoosingLock(AcceptanceTester $I): void
    {
        $this->setConfig('keepAliveInterval', 1);

        $I->amOnPage('/admin');

        $this->loginAuthor1($I);
        $I->waitForText('You have been successfully logged in');

        $I->amOnPage('/admin/pages/home');
        $I->see('Home');

        $locks = $this->get_LOCKFILE();

        assertCount(1, $locks);
        assertArrayHasKey('/admin/pages/home', $locks);

        // Remove lock as if it has expired
        $this->set_LOCKFILE([]);

        $I->wait(3);

        $locks = $this->get_LOCKFILE();

        assertCount(0, $locks);
        assertArrayNotHasKey('/admin/pages/home', $locks);

        $I->see('Lock on page has expired or has been removed.');
    }

    private function loginAuthor1(AcceptanceTester $I): void
    {
        $I->fillField('data[username]', 'author1');
        $I->fillField('data[password]', 'Page@Locks0');

        $I->click('.button.primary');
    }

    private function set_LOCKFILE(array $locks): void
    {
        $this->lockFile->save($locks);
        $this->lockFile->free();
    }

    private function get_LOCKFILE(): array
    {
        $locks = $this->lockFile->content();
        $this->lockFile->free();

        return $locks;
    }

    protected function setConfigDefaults(): void
    {
        $this->configFile->save($this->configDefaults);
        $this->configFile->free();
    }

    /**
     * @param string $key
     * @param int|bool $value
     */
    protected function setConfig(string $key, $value): void
    {
        $defaults = $this->configDefaults;
        $defaults[$key] = $value;

        $this->configFile->save($defaults);
        $this->configFile->free();
    }
}


pagelocks.php
<?php

namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Exception;
use Grav\Common\Assets;
use Grav\Common\Plugin;
use Grav\Plugin\PageLocks\LockHandler;

/**
 * Class PageLocksPlugin
 * @package Grav\Plugin
 */
class PageLocksPlugin extends Plugin
{
    protected string $route = 'locks';

    /**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     *     that the plugin wants to listen to. The key of each
     *     array section is the event that the plugin listens to
     *     and the value (in the form of an array) contains the
     *     callable (or function) as well as the priority. The
     *     higher the number the higher the priority.
     */
    public static function getSubscribedEvents(): array
    {
        return [
            'onPluginsInitialized' => [
                // Uncomment following line when plugin requires Grav < 1.7
                // ['autoload', 100000],
                ['onPluginsInitialized', 0]
            ]
        ];
    }

    /**
     * Composer autoload
     *
     * @return ClassLoader
     */
    public function autoload(): ClassLoader
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    /**
     * Initialize the plugin
     */
    public function onPluginsInitialized(): void
    {
        if ($this->isAdmin()) {
            // Handle possible PageLocks requests
            $lockHandler = new LockHandler();
            $response = $lockHandler->handleRequest();

            // if there was an async PageLocks request, return response and stop processing.
            if ($response) {
                echo json_encode($response);
                die();
            }

            $this->enable([
                // Put your main events here
                'onAssetsInitialized' => ['onAssetsInitialized', 0],
                'onAdminMenu' => ['onAdminMenu', 0],
                'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
            ]);

            return;
        }

        // Enable the main events we are interested in
        $this->enable([
            // Put your main events here
        ]);
    }

    /**
     * Add assets required for page
     */
    public function onAssetsInitialized(): void
    {
        if (!$this->config) {
            throw new Exception('Property "$this->config" should not be null.');
        }

        // Should minified assets be used?
        $min = $this->config->get('plugins.pagelocks.productionMode', true) ? '.min' : '';

        /** @var Assets */
        $assets = $this->grav['assets'];

        // Add script for all Admin pages. Must at least check on which page user is.
        $assets->addJs("plugin://pagelocks/js/pagelocker$min.js");
        $assets->addCss("plugin://pagelocks/css/page$min.css");

        // Add scripts required for Admin page of PageLocks:
        // ends with $this->config['plugins']['admin']['route']/locks
        $route = $this->grav['uri']->uri();
        $pagelocksadmin = $this->config->get('plugins.admin.route', '/admin') . "/locks";

        // dump($this->config['plugins']['admin']['route']);
        // if (preg_match('/\/admin\/locks$/', $route) === 1) {
        if (strpos($route, $pagelocksadmin) !== false) {
            $assets->addJs("plugin://pagelocks/js/pagelocksadmin$min.js");
            $assets->addCss("plugin://pagelocks/css/lock-admin$min.css");
        }
    }

    /**
     * Add navigation item to the admin plugin
     */
    public function onAdminMenu(): void
    {
        $this->grav['twig']->plugins_hooked_nav['PLUGIN_PAGELOCKS.LOCKS'] = [
            'route' => $this->route,
            'icon' => 'fa-lock'
        ];
    }

    /**
     * Add plugin templates path
     */
    public function onTwigTemplatePaths(): void
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/admin/templates';
    }
}


classes/LockHandler.php
<?php

namespace Grav\Plugin\PageLocks;

use DateTime;
use Exception;
use Grav\Common\Data\Data;
use Grav\Common\Grav;
use Grav\Common\User\DataUser\User;
use RocketTheme\Toolbox\File\File;
use RocketTheme\Toolbox\ResourceLocator\UniformResourceLocator;

class LockHandler
{
    const LOGFILE = 'user-data://pagelocks/debug.log';

    protected Grav $grav;
    protected Data $config;
    protected Storage $storage;
    protected bool $debug;
    protected string $pathToLogFile;

    public function __construct()
    {
        $this->grav = Grav::instance();
        $this->config = $this->grav['config'];
        $this->storage = new Storage();
        $this->debug = $this->config->get('plugins.pagelocks.debug', true);
        $this->pathToLogFile = $this->getPathToLogFile();
    }

    /**
     * Handle async requests send by javascript injected by PageLocks into Admin panel.
     * 
     * @return ?array The response specific to each type of request, or null if no PageLocks request.
     */
    public function handleRequest(): ?array
    {
        if ($this->isNotSendByPageLocks()) {
            return null;
        }

        if ($this->isReadLocks()) {
            $locks = $this->readLocks();
            return $locks;
        }

        /** @var User */
        $user = $this->grav['user'];

        $locks = $this->storage->readLocksForUpdate();

        if ($this->isAcquireLock()) {
            $response = $this->acquireLock($locks, $user);
        } elseif ($this->isKeepAlive()) {
            $response = $this->keepAlive($locks);
        } elseif ($this->isForceRemoveLock()) {
            $response = $this->forceRemoveLock($locks, $user);
        } else {
            throw new Exception('Should not reach this.');
        }

        $this->clearExpiredLocks($locks);
        $this->storage->saveLocks($locks);

        return $response;
    }

    /**
     * Check if request originated from front-end javascript from PageLocks.
     * 
     * @return bool True if request is not from PageLocks
     */
    protected function isNotSendByPageLocks(): bool
    {
        return !(
            $this->isKeepAlive() ||
            $this->isAcquireLock() ||
            $this->isReadLocks() ||
            $this->isForceRemoveLock()
        );
    }

    /**
     * Check if request is a 'readLocks' request.
     * 
     * @return bool True if request is a 'readLocks' request, else false.
     */
    protected function isReadLocks(): bool
    {
        return isset($_POST['readLocks']);
    }

    /**
     * Check if request is a 'acquireLock' request.
     * 
     * @return bool True if request is a 'acquireLock' request, else false.
     */
    protected function isAcquireLock(): bool
    {
        return isset($_POST['acquireLock']);
    }

    /**
     * Check if request is a 'keepAlive' request.
     * 
     * @return bool True if request is a 'keepAlive' request, else false.
     */
    protected function isKeepAlive(): bool
    {
        return isset($_POST['keepAlive']);
    }

    /**
     * Check if request is a 'forceRemoveLocks' request.
     * 
     * @return bool True if request is a 'forceRemoveLocks' request, else false.
     */
    protected function isForceRemoveLock(): bool
    {
        return isset($_POST['forceRemoveLock']);
    }

    /**
     * Get list of locks for PageLocks admin page.
     * 
     * @return array Response for 'readLocks' request.
     */
    protected function readLocks(): array
    {
        if ($this->debug) $this->log("/admin/locks : Read all locks\n");

        $locks = $this->storage->readLocks();
        
        return [
            'locks' => $locks->toArray(),
            'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.CONFIRM_DELETE'),
            'countAlert' => $this->translate(['PLUGIN_PAGELOCKS.ALERT.LOCK_COUNT', count($locks)]),
        ];
    }

    /**
     * Set lock if page is not already locked by other user.
     * 
     * @return array Response for 'acquireLock' request.
     */
    protected function acquireLock(Locks $locks, User $user): array
    {
        $route = $this->filteredPOST('acquireLock');

        // Remove existing lock for user
        $this->removeLockOfUser($locks, $user);

        $keepAliveInterval = $this->config->get('plugins.pagelocks.keepAliveInterval', 60) * 1000;

        if (!$this->userIsOnPage($route)) {
            return [
                'isOnPage' => false,
                'isLockGranted' => false,
                'byUser' => $user->fullname,
                'keepAliveInterval' => $keepAliveInterval,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_REMOVED')
            ];
        }

        // Acquire new lock for new page

        $userHoldingLock = $this->isPageLocked($locks, $route, $user);

        if ($userHoldingLock) {
            // Page has already been locked by other user
            if ($this->debug) $this->log("$route : Lock request by $user->email, already locked by $userHoldingLock\n");

            return [
                'isOnPage' => true,
                'isLockGranted' => false,
                'byUser' => $userHoldingLock,
                'keepAliveInterval' => $keepAliveInterval,
                'alert' => $this->translate([
                    'PLUGIN_PAGELOCKS.ALERT.ALREADY_LOCKED',
                    $userHoldingLock, $userHoldingLock
                ])
            ];
        } else {
            // Lock has been granted
            if ($this->debug) $this->log("$route : Lock acquired: By $user->email\n");

            $this->setLock($locks, $route, $user);
            return [
                'isOnPage' => true,
                'isLockGranted' => true,
                'byUser' => $user->fullname,
                'keepAliveInterval' => $keepAliveInterval,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_GRANTED'),
            ];
        }
    }

    /**
     * Check if user is on a Page in Admin.
     * 
     * @return bool True if user is on /admin/pages/<path-to-page>
     */
    public function userIsOnPage(string $route): bool
    {
        return preg_match('/\/pages(\/[a-z][a-z0-9\-_]*)+$/', $route) === 1;
    }

    /**
     * Update timestamp of lock to extend the lifetime of the lock.
     * 
     * @return array Response for 'keepAlive' request.
     */
    protected function keepAlive(Locks $locks): array
    {
        $route = $this->filteredPOST('keepAlive');

        if (isset($locks[$route])) {
            if ($this->debug) $this->log("$route : Lock extended: Held by {$locks[$route]->email}\n");

            $now = (new DateTime())->getTimestamp();
            $locks[$route]->timestamp = $now;

            return [
                'isExtended' => true,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_EXTENDED'),
            ];
        } else {
            return [
                'isExtended' => false,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_EXPIRED'),
            ];
        }
    }

    /**
     * Remove lock held on page.
     * 
     * @return array Response for 'forceRemoveLocks' request.
     */
    protected function forceRemoveLock(Locks $locks, User $user): array
    {
        $route = $this->filteredPOST('forceRemoveLock');

        if (isset($locks[$route])) {
            if ($this->debug) $this->log("$route : Lock removed by force by $user->email\n");

            unset($locks[$route]);

            return [
                'isLockRemoved' => true,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_REMOVED')
            ];
        } else {
            return [
                'isLockRemoved' => false,
                'alert' => $this->translate('PLUGIN_PAGELOCKS.ALERT.LOCK_NOT_REMOVED')
            ];
        }
    }

    /**
     * Remove lock held by user.
     */
    protected function removeLockOfUser(Locks $locks, User $user): void
    {
        foreach ($locks as $route => $lock) {
            if ($lock->email === $user->email) {
                if ($this->debug) $this->log("$route : Lock removed: Was held by {$locks[$route]->email}\n");

                unset($locks[$route]);
            }
        }
    }

    /**
     * Set lock on page for user.
     */
    protected function setLock(Locks $locks, string $route, User $user): void
    {
        $now = (new DateTime())->getTimestamp();

        $locks[$route] = new Lock([
            'email' => $user->email,
            'fullname' => $user->fullname,
            'timestamp' => $now,
        ]);
    }

    /**
     * Check if page is locked by other user
     * 
     * @return string Email address of user locking the page, or '' if no lock.
     */
    protected function isPageLocked(Locks $locks, string $route, User $user): string
    {
        if (isset($locks[$route])) {
            $lock = $locks[$route];

            if ($lock->email != $user->email) {
                return $lock->email;
            }
        }

        return '';
    }

    /**
     * Loops through all locks and removes expired locks
     */
    private function clearExpiredLocks(Locks $locks): void
    {
        $expiresAfter = $this->config->get('plugins.pagelocks.expiresAfter', 3600);
        $now = (new DateTime())->getTimestamp();

        foreach ($locks as $route => $lock) {
            $lastKeepalive = $lock->timestamp;
            $interval = $now - $lastKeepalive;

            if ($interval > $expiresAfter) {
                if ($this->debug) $this->log("$route : Clear expired lock\n");

                unset($locks[$route]);
            }
        }
    }

    /**
     * Sanitize data received via $_POST.
     * 
     * @param string $key The value to be sanitized.
     * @return string Sanitized data.
     */
    protected function filteredPOST(string $key): string
    {
        switch ($key) {
            case 'keepAlive':
            case 'acquireLock':
            case 'forceRemoveLock':
                return filter_var($_POST[$key], FILTER_SANITIZE_URL);
            default:
                throw new Exception("Case '$key' is not being handled.");
        }

        return '';
    }

    /**
     * Log data in log file for debugging purposes.
     */
    protected function log(string $message): void
    {
        /** @var File */
        $file = File::instance($this->pathToLogFile);
        $file->lock();

        $content = $file->content();
        $now = date('c', time());

        $content = "$now\t$message$content";
        
        $file->save($content);
        $file->unlock();
    }

    /**
     * Get the path to the debug.log file.
     * 
     * @return string The path to the debug.log file
     * @throws Exception When path to log file cannot be found.
     */
    protected function getPathToLogFile(): string {
        /** @var UniformResourceLocator */
        $locator = $this->grav['locator'];
        $path = $locator->findResource(self::LOGFILE, true, true);

        if ($path === false) {
            throw new Exception('Path for log file cannot be found');
        }

        return $path;
    }

    /**
     * Translate alerts
     * 
     * @param string|array $args Can be a string, or an array containing a string and its parameters.
     */
    protected function translate($args): string
    {
        return $this->grav['language']->translate($args);
    }
}
